io/io/inc/ROOT/RFile.hxx

```diff
@@ -0,0 +1,147 @@
+/// \file ROOT/RFile.hxx
+/// \ingroup Base ROOT7
+/// \author Giacomo Parolini <giacomo.parolini@cern.ch>
+/// \date 2025-03-19
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+#ifndef ROOT7_RFile
+#define ROOT7_RFile
+
+#include <ROOT/RError.hxx>
+
+#include <TClass.h>
+
+#include <string_view>
+#include <memory>
+
+class TKey;
+class TFile;
+
+namespace ROOT {
+namespace Experimental {
+
+class RFile;
+struct RFileKeyInfo;
+
+namespace Internal {
+
+ROOT::RLogChannel &RFileLog();
+
+} // namespace Internal
+
+/**
+\class ROOT::Experimental::RFile
+\ingroup RFile
+\brief An interface to read from, or write to, a ROOT file, as well as performing other common operations.
+
+TODO: more in-depth explanation
+*/
+class RFile final {
+   enum PutFlags {
+      kPutAllowOverwrite = 0x1,
+      kPutOverwriteKeepCycle = 0x2,
+   };
+
+   std::unique_ptr<TFile> fFile;
+
+   // Outlined to avoid including TFile.h
+   explicit RFile(std::unique_ptr<TFile> file);
+
+   /// Gets object `path` from the file and returns an **owning** pointer to it.
+   /// The caller should immediately wrap it into a unique_ptr of the type described by `type`.
+   [[nodiscard]] void *GetUntyped(std::string_view path, const TClass &type) const;
+
+   /// Writes `obj` to file, without taking its ownership.
+   void PutUntyped(std::string_view path, const TClass &type, const void *obj, std::uint32_t flags);
+
+   /// \see Put
+   template <typename T>
+   void PutInternal(std::string_view path, const T &obj, std::uint32_t flags)
+   {
+      const TClass *cls = TClass::GetClass(typeid(T));
+      if (!cls) {
+         throw ROOT::RException(R__FAIL(std::string("Could not determine type of object ") + std::string(path)));
+      }
+      PutUntyped(path, *cls, &obj, flags);
```

Comment:
In principle, we could even avoid a dependency of the other on `TClass` by only forwarding the result of `typeid(T)` and call `TClass::GetClass` only in `RFile.cxx`. Not sure if this buys us anything...